drivers/qmc5883l/qmc5883l.c

```diff
@@ -0,0 +1,203 @@
+ *
+ * @}
+ */
+
+#include <limits.h>
+
+#include "assert.h"
+#include "periph/gpio.h"
+#include "qmc5883l.h"
+#include "qmc5883l_internal.h"
+
+/* shortcut to the i2c device address */
+#define ADDR        QMC5883L_ADDR
+
+static int _reg_read(const qmc5883l_t *dev, uint8_t reg,
+                     uint8_t *val, int acquire, int release)
+{
+    if (acquire) {
+        if (i2c_acquire(dev->i2c) != 0) {
+            return QMC5883L_BUSERR;
+        }
+    }
+    int res = i2c_read_reg(dev->i2c, ADDR, reg, val, 0);
+    if ((release) || (res != 0)) {
+        i2c_release(dev->i2c);
+    }
+    return (res == 0) ? QMC5883L_OK : QMC5883L_BUSERR;
+}
+
+static int _reg_write(const qmc5883l_t *dev, uint8_t reg,
+                      uint8_t val, int acquire, int release)
+{
+    if (acquire) {
+        if (i2c_acquire(dev->i2c) != 0) {
+            return QMC5883L_BUSERR;
+        }
+    }
+    int res = i2c_write_reg(dev->i2c, ADDR, reg, val, 0);
+    if ((release) || (res != 0)) {
+        i2c_release(dev->i2c);
+    }
+    return (res == 0) ? QMC5883L_OK : QMC5883L_BUSERR;
+}
+
+int qmc5883l_init(qmc5883l_t *dev, const qmc5883l_params_t *params)
+{
+    assert(dev);
+    assert(params);
+    int res;
+
+    dev->i2c = params->i2c;
+    dev->pin_drdy = params->pin_drdy;
+    dev->cfg = (params->odr | params->rng | params->osr | QMC5883L_CONT);
+
+    /* lets start with a soft reset */
+    res = _reg_write(dev, QMC5883L_CTRL2, QMC5883L_SOFT_RST, 1, 0);
+    if (res != QMC5883L_OK) {
+        return res;
+    }
+
+    /* verify the reset by reading the status register, should be all zeros */
+    uint8_t tmp;
+    if (i2c_read_reg(dev->i2c, ADDR, QMC5883L_STATUS, &tmp, 0) != 0) {
+        i2c_release(dev->i2c);
+        return QMC5883L_BUSERR;
+    }
+    if (tmp != 0) {
+        i2c_release(dev->i2c);
+        return QMC5883L_NOCFG;
+    }
+
+    /* write the actual device configuration */
+    res = _reg_write(dev, QMC5883L_SETRESET, 0x01, 0, 0);
+    if (res != QMC5883L_OK) {
+        return res;
+    }
+    res = _reg_write(dev, QMC5883L_CTRL2, QMC5883L_INT_ENB, 0, 0);
+    if (res != QMC5883L_OK) {
+        return res;
+    }
+    return _reg_write(dev, QMC5883L_CTRL1, (dev->cfg | QMC5883L_CONT), 0, 1);
+}
+
+int qmc5883l_data_ready(const qmc5883l_t *dev)
+{
+    assert(dev);
+    uint8_t status;
+
+    int res = _reg_read(dev, QMC5883L_STATUS, &status, 1, 1);
+    if (res != QMC5883L_OK) {
+        return res;
+    }
+    return (status & QMC5883L_DRDY) ? QMC5883L_OK : QMC5883L_NODATA;
+}
+
+int qmc5883l_read(const qmc5883l_t *dev, int16_t *data_out)
+{
+    assert(data_out);
+    int16_t tmp[3];
+
+    int res = qmc5883l_read_raw(dev, tmp);
+    if ((res == QMC5883L_OK) || (res == QMC5883L_OVERFLOW)) {
+        unsigned scale = (dev->cfg & QMC5883L_RNG_8G) ? 8000 : 2000;
+        for (unsigned i = 0; i < 3; i++) {
+            data_out[i] = (int16_t)((((int32_t)tmp[i]) * scale) / INT16_MAX);
```

Comment:
I'm not sure whether this conversion is correct. According to the datasheet the sensitivity of the sensor is:

| RNG | Typical | Unit |
| ------- | -------- | ----- |
| Field Range = ±2G | 12000 | LSB/Gs |
| Field Range = ±8G | 3000 | LSB/Gs |

That is, the value in mGs would be given by `tmp[i] / 12` and `tmp[i] / 3`, respectively.

Reply:
Hm, possible. The datasheet seems not too concise on this topic: in the data register description it quotes `Each axis has 16 bit data width in 2’s complement,`. Using the values above, it would mean the actual possible ranges would be `+-2.6G / +-10.6G` when using the full 16-bit resolution. This seems strange to me, as all other magenometers I know map their G range to their full analog resolution, so I went with that.

Reply:
IMHO, Unit/LSB always gives the resolution of the values in output registers. The datasheet for HMC5883L is probably a bit more clear, it says:

_Use the “Gain” column to convert counts to Gauss. The “Digital Resolution” column is the theoretical value in term of milli-Gauss per count (LSb) which is the inverse of the values in the “Gain” column."_

And the tables looks like:

Recommended Sensor Field Range | Gain (LSb/ Gauss) | Digital Resolution (mG/LSb) | Output Range |
| -------------:| ------:| ------:| ------------------------------------------:|
| ± 0.88 Ga | 1370 | 0.73 | 0xF800–0x07FF (-2048–2047 ) |
| ± 1.3 Ga | 1090 (default) | 0.92 | 0xF800–0x07FF (-2048–2047 ) |
| ± 1.9 Ga | 820 | 1.22 | 0xF800–0x07FF (-2048–2047 ) |
| ± 2.5 Ga | 660 | 1.52 | 0xF800–0x07FF (-2048–2047 ) |
| ± 4.0 Ga | 440 | 2.27 | 0xF800–0x07FF (-2048–2047 ) |
| ± 4.7 Ga | 390 | 2.56 | 0xF800–0x07FF (-2048–2047 ) |
| ± 5.6 Ga | 330 | 3.03 | 0xF800–0x07FF (-2048–2047 ) |
| ± 8.1 Ga | 230 | 4.35 | 0xF800–0x07FF (-2048–2047 ) |

The full-scale range in data sheet is given by ± 8 Gs even though it can be ± 8.1.

Reply:
Ok, lets go with that. Will adjust the driver accordingly